### ISSUE TITLE
Redirect logins to originating page

### ISSUE THREAD

**Issue by sboligitz** [rubyforgood/circulate#216]
As a CTL member not currently logged in, when clicking on "Member Login" or a login link from a page (e.g., an item page), I want to be redirected to the page I came from after a successful login, so that I can resume browsing where I left off. (Per @avinashupadhya99 's suggestion.)

RK note: see also #174 .

RK additional note: situations that we haven't fully worked out yet:
- if someone is going the Login page from outside the app
- if someone goes to the Login page from a direct URL (e.g., types https://circulate-staging.herokuapp.com/users/sign_in into their browser bar)
- if someone attempts to access a page that can only be accessed when logged in

**Comment by avinashupadhya99** [comment]
Hey, can I work on this? If yes, please do assign this to me. Thank you.

**Comment by avinashupadhya99** [comment]
Hey! I was just curious how do we know if an user who is not logged in is a CTL member? Currently there is a `Login/SignUp` option for non-logged in users in place of `Hold`. I can implement a feature where users are redirected back to the previous url after logging in. Please let me know your thoughts about this.

**Comment by rkim00** [comment]
Hi @avinashupadhya99 , thanks for the question. I think that your implementation generally makes sense.

Note that there's currently a "Member Login" link in the upper-right. Currently, after clicking on this link, a member goes to https://circulate-staging.herokuapp.com/, which is incorrect; users who are just members should go to the My Loans page, and staff should go to the admin page (#174 is meant to fix this) ). You don't need to implement #174 (someone else is assigned to that), but whatever you do should largely leave that functionality the same.

So if you click on the "login" link in the listing for the item, and then sign-in, you should be redirected to the same item page. But if you click on the "Member Login" link in the upper-right, you should go to the My Loans page (or as currently wrongly implemented, to the https://circulate-staging.herokuapp.com/ ).

Does that all make sense? I'm not sure of the best way to implement that, but give it a shot. Let me know if you have any questions.

**Comment by avinashupadhya99** [comment]
Yes got it. Thanks.

**Comment by avinashupadhya99** [comment]
Hey @rkim00, I was thinking if we can redirect users to their previous page unless that previous page was the home page i.e, [ https://circulate-staging.herokuapp.com/ ]( https://circulate-staging.herokuapp.com/ ). If the user logs in/ signs up through the home page's `Log In/ Sign Up`, then we can redirect them to the respective pages which the issue #174 addresses.
I am thinking about implementing this approach. I do have another complicated approach which uses HTML5's `Local Storage` to store the origin of the `Log In/Sign Up` request and then redirect in the controller based on the origin. This is too complicated. 
Please let me know your thoughts. Also, please ask if I was not clear.

**Comment by rkim00** [comment]
Thanks, great thinking on this. I have been contemplating this a little bit, and I think you're suggesting something different from what I proposed, but I think your suggestion is better! :)

So you're proposing that clicking on "Member Login" redirects the user to wherever they originated from; this makes sense to me. We might as well do the same thing in the home page case; I don't see any reason to make a distinction.

I re-wrote the issue description to match all this.

But I see a few more situations that we _should_ think about:
1) if someone is coming to the Login page from outside the app
2) if someone goes to the Login page from a direct URL (e.g., the user types `https://circulate-staging.herokuapp.com/users/sign_in` into their browser bar)
3) if someone attempts to access a page that can only be accessed when logged in ( `https://circulate-staging.herokuapp.com/admin` )

(3) above currently gives a 500 error, which isn't a great user experience, but maybe is okay enough for the moment.

For (1) or (2) above, I think we should redirect as #174 mentions (based whether the user is librarian/staff/admin or just a member). How does that sound? Is that easy or hard to do?

**Comment by avinashupadhya99** [comment]
I think we should redirect users to `/admin` or `/loans` if the request comes from the home page and also in cases of (1) and (2). If we are **not** redirecting to `/admin` or `/loans`, the issue #174 would be in vain. 
The information regarding the previous page would be in `request.referer`. 
Also I think we should open a new issue for the scneario (3). I would be willing to look into that. You can assign me to that. 
If you think we should not make any distinction from where the request came from and redirect users to their previous page, I can still do that. 
Please let me know and I will implement this at the earliest. Thanks.

**Comment by rkim00** [comment]
Hi @avinashupadhya99 . Thanks for all your work on this!

So regarding redirecting to /admin or /loans from the home page; I think that would be fine, yes. The reality, however, is that people don't come to the app home page at the moment. Chicago Tool Library uses a SquareSpace page (https://chicagotoollibrary.org/) that links into the different parts of the app. Clicking on the log in the app (at least in the staging version that we're working on), redirects to this SquareSpace page ( https://chicagotoollibrary.org/ ), not https://circulate-staging.herokuapp.com/ . People would only see the "app home page" if they navigate directly there.

I think redirecting to /admin or /loans from the home page is probably a better experience, and possibly when Circulate is used by other organizations they may use this page more than Chicago Tool Library does. But I didn't want you to write more code for a situation that doesn't really show up now.

Basically, feel free to implement whichever solution you like for the home page.

For (1) and (2) above, we should definitely redirect to /admin or /loans .

I'll open a new issue for (3) and assign to you.

Thank you very much!

**Comment by avinashupadhya99** [comment]
Great, I will implement it soon.
